FIX: round instead of truncate Agg canvas size

tacaswell opened the pull request:
As suggested by @njsmith  in #8253 round, rather than truncate the size in pixels of the Agg canvas.  This fails two tests for me locally

 - test_boxarrow[0-boxarrow_test_image-png] 
 - test_bbox_inches_tight_suptile_legend[0-bbox_inches_tight_suptile_legend-png]

iirc the second one is also unstable on travis.  Do not have time to investigate further tonight, but regenerating two images seems like a small price to pay for making it more reliable to get renders of a fixed pixel size.

tacaswell committed:
FIX: round instead of truncate Agg canvas size

efiring commented on lib/matplotlib/backends/backend_webagg_core.py:
```diff
@@ -243,7 +243,7 @@ def get_renderer(self, cleared=None):
         # in get_diff_image
-        _, _, w, h = self.figure.bbox.bounds
+        _, _, w, h = np.round(self.figure.bbox.bounds).astype(int)
         w, h = int(w), int(h)
```
This conversion to python int is not done in the other two backends, so presumably it is not needed here.  (I see that numpy.int64 and int instances both hash to the corresponding python int.)

tacaswell committed:
MNT: if the figure size is not changed, bail early

This avoids marking the figure as stale un-necessarily.

Also remove a local variable which is immediately replaced below.

tacaswell committed:
API: when forwarding figure resize, round not truncate pixel numbers

tacaswell committed:
TST: regenerate a test image

The change to rounding made these images one pixel wider

tacaswell committed:
API: feedback to figure size due to pixel counts

When rasterizing the figure, the allowed sizes are discrete due to a
finite dpi.  When getting the renderer (at which point we have
committed to rasterizing the figure at this dpi) feedback the actual
size.

tacaswell commented:
I get one failure locally now:

![bbox_inches_tight_suptile_legend-failed-diff](https://cloud.githubusercontent.com/assets/199813/23827774/8182bed8-0689-11e7-91be-29edaa5e5fe3.png)

which is due to some text shifting vertically, not really sure what is going on here.  The logic inside of `bbox_tight` is a bit confusing and hacks at the inneards of the transform stack in interesting ways.

tacaswell committed:
API: add API change note

ngoldbaum commented:
FWIW, this doesn't fix the floating point round-off issues in the animation framework that #8253 resolves. I will try to make a simple test script to attach to that issue.

efiring commented:
@ngoldbaum, is it possible that using rounding consistently--here and in #8253, in place of int with the nextafter adjustment--would solve the problem?

ngoldbaum commented:
I don't know - all I tried was running my test yt script with this PR applied in my working copy. It failed, but it's possible there needs to be other code changes in the animation module.

jklymak commented:
Thanks for the PR. This hasn't seen action for a long time so marking stale.  Note that PRs sometimes fall through the cracks, so if this is important, please ping for more reviews.

tacaswell commented:
```
self.window.ClientToWindowSize(wx.Size(\nRuntimeError: wrapped C/C++ object of type FigureFrameWx has been deleted
```

This still needs some work and I think tweak around the short-circuit on the resize is not working quite right either.

efiring commented:
It seems past time for something like this to get in; let's keep it alive.

QuLogic commented:
Looking at the results here, we have:
- Contains text:
    - FAILED lib/matplotlib/tests/test_arrow_patches.py::test_boxarrow[png] - matplotlib.testing.exceptions.ImageComparisonFailure: Image sizes do not match expected size: (833, 266, 3) actual size (833, 267, 3)
    - FAILED lib/mpl_toolkits/axes_grid1/tests/test_axes_grid1.py::test_image_grid_single_bottom[png] - matplotlib.testing.exceptions.ImageComparisonFailure: Image sizes do not match expected size: (166, 288, 3) actual size (167, 289, 3)
- No text in image:
    - FAILED lib/mpl_toolkits/axes_grid1/tests/test_axes_grid1.py::test_image_grid[png] - matplotlib.testing.exceptions.ImageComparisonFailure: Image sizes do not match expected size: (332, 332, 3) actual size (333, 333, 3)
- Not actually images:
    - FAILED lib/matplotlib/tests/test_offsetbox.py::test_annotationbbox_extents[window_extent] - AssertionError: 
    - FAILED lib/matplotlib/tests/test_offsetbox.py::test_annotationbbox_extents[tightbbox] - AssertionError: 

So there's a minor case for putting these into the text overhaul, but it doesn't seem super important, and could be done by itself.

ayshih commented:
#32038 has now fixed the bug with the canvas height due to a floating-point issue without switching to rounding "true" fractional pixels (e.g., 0.75 of a pixel).  I think that rounding such "true" fractional pixels may have some undesired effects when rendering artists that are supposed to run up to the edge of the canvas.  In the absence of an explicit request to change how "true" fractional pixels" are handled, I suggest closing this PR.
